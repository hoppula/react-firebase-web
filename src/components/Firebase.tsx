import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

export interface FirebaseProps {
  readonly apiKey: string
  readonly projectId: string
  readonly databaseURL?: string
}

export class Firebase extends React.Component<FirebaseProps, {}> {
  static propTypes = {
    apiKey: PropTypes.string.isRequired,
    databaseURL: PropTypes.string,
    projectId: PropTypes.string.isRequired
  }

  static childContextTypes = {
    firebase: PropTypes.object
  }

  firebase: firebase.app.App

  constructor(props: FirebaseProps) {
    super(props)
    const { apiKey, projectId, databaseURL } = props
    this.firebase = firebase.initializeApp({
      apiKey,
      authDomain: `${projectId}.firebaseapp.com`,
      databaseURL: databaseURL
        ? databaseURL
        : `https://${projectId}.firebaseio.com`,
      storageBucket: `${projectId}.appspot.com`
    })
  }

  getChildContext() {
    return {
      firebase: this.firebase
    }
  }

  render() {
    return React.Children.only(this.props.children)
  }
}

export default Firebase
