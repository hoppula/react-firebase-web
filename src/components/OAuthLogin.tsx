import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"
import { FirebaseContext } from "../types"

export interface AuthProviders {
  facebook: "FacebookAuthProvider"
  github: "GithubAuthProvider"
  google: "GoogleAuthProvider"
  twitter: "TwitterAuthProvider"
}

export interface AuthFlows {
  popup: "signInWithPopup"
  redirect: "signInWithRedirect"
}

const authProviders: AuthProviders = {
  facebook: "FacebookAuthProvider",
  github: "GithubAuthProvider",
  google: "GoogleAuthProvider",
  twitter: "TwitterAuthProvider"
}

const authFlows: AuthFlows = {
  popup: "signInWithPopup",
  redirect: "signInWithRedirect"
}

export interface OAuthProps {
  readonly flow?: "popup" | "redirect"
  readonly provider: "facebook" | "google" | "twitter" | "github"
  readonly children: (
    authenticate: () => Promise<any>
  ) => React.ReactElement<any>
  readonly scopes?: string[]
}

export class OAuthLogin extends React.Component<OAuthProps, {}> {
  static propTypes = {
    flow: PropTypes.oneOf(Object.keys(authFlows)),
    provider: PropTypes.oneOf(Object.keys(authProviders)).isRequired,
    children: PropTypes.func.isRequired,
    scopes: PropTypes.array
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  static defaultProps = {
    flow: "popup"
  }

  context: FirebaseContext

  authenticate = () => {
    const { flow, provider, scopes = [] } = this.props
    const authProvider = new firebase.auth[authProviders[provider]]()
    if (
      authProvider instanceof firebase.auth.FacebookAuthProvider ||
      authProvider instanceof firebase.auth.GithubAuthProvider ||
      authProvider instanceof firebase.auth.GoogleAuthProvider
    ) {
      scopes.forEach(scope => {
        authProvider.addScope(scope)
      })
    }
    return this.context.firebase.auth()[authFlows[flow]](authProvider)
  }

  render() {
    return React.Children.only(this.props.children(this.authenticate))
  }
}

export default OAuthLogin
