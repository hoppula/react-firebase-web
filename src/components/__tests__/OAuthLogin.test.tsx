import React from "react"
import renderer from "react-test-renderer"
import firebase from "firebase/app"
import { Firebase } from "../Firebase"
import OAuthLogin from "../OAuthLogin"

import { initializeMockSDK } from "./setup"
const mocksdk = initializeMockSDK()

jest.mock("firebase/app", () => {
  return {
    initializeApp: (...args: any[]) => mocksdk.initializeApp(...args)
  }
})
// override firebase.auth with mock
Object.defineProperty(firebase, "auth", { value: mocksdk.auth })

describe("OAuthLogin", () => {
  test("login sets auth data properly", () => {
    let component = renderer.create(
      <Firebase apiKey="firebase-test" projectId="firebase-test">
        <OAuthLogin
          flow="popup"
          provider="google"
          children={login => <button onClick={login}>Login with Google</button>}
        />
      </Firebase>
    )
    const auth = component.getInstance().firebase.auth()
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(auth.currentUser).toBe(null)

    const button = component.root.findByType("button")
    button.props.onClick()
    auth.flush()

    expect(auth.currentUser.providerData[0].providerId).toBe("google.com")
  })
})
