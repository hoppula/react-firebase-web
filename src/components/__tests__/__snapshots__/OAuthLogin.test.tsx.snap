// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuthLogin login sets auth data properly 1`] = `
<button
  onClick={[Function]}
>
  Login with Google
</button>
`;
